source/opt/ir_context.h

```diff
@@ -121,8 +132,64 @@ class IRContext {
   // Appends a function to this module.
   inline void AddFunction(std::unique_ptr<Function>&& f);
 
+  // Returns a pointer to a def-use manager.  If the def-use manager is
+  // invalid, it is rebuilt first.
+  opt::analysis::DefUseManager* get_def_use_mgr() {
+    if (!AreAnalysesValid(kAnalysisDefUse)) BuildDefUseManager();
+    return def_use_mgr_.get();
+  }
+
+  // Builds the def-use manager from scratch, even if it was already valid.
+  void BuildDefUseManager() {
+    def_use_mgr_.reset(new opt::analysis::DefUseManager(consumer(), module()));
+    valid_analyses_ = (Analysis)(valid_analyses_ | kAnalysisDefUse);
+  }
+
+  // Sets the message consumer to the given |consumer|. |consumer| which will be
+  // invoked every time there is a message to be communicated to the outside.
+  void SetMessageConsumer(spvtools::MessageConsumer c) {
+    consumer_ = std::move(c);
+  }
+
+  // Returns the reference to the message consumer for this pass.
+  const spvtools::MessageConsumer& consumer() const { return consumer_; }
+
+  // Rebuilds the analyses in |set| that are invalid.
+  void BuildAnalyses(Analysis set);
+
+  // Invalidates all of the analyses except for those in |preserved_analyses|.
+  void InvalidateAnalyses(Analysis preserved_analyses);
+
+  // Returns true if all of the valid analyses are consistent with module.
+  bool Consistent() { return true; }
+
+  // Turns the instruction defining the given |id| into a Nop. Returns true on
+  // success, false if the given |id| is not defined at all. This method also
+  // erases both the uses of |id| and the information of this |id|-generating
+  // instruction's uses of its operands.
+  bool KillDef(uint32_t id);
+
+  // Turns the given instruction |inst| to a Nop. This method erases the
+  // information of the given instruction's uses of its operands. If |inst|
+  // defines an result id, the uses of the result id will also be erased.
+  void KillInst(ir::Instruction* inst);
+
+  // Returns true if all of the given analyses are valid.
+  bool AreAnalysesValid(Analysis set) { return (set & valid_analyses_) == set; }
+
+  // Replaces all uses of |before| id with |after| id. Returns true if any
+  // replacement happens. This method does not kill the definition of the
+  // |before| id. If |after| is the same as |before|, does nothing and returns
+  // false.
+  bool ReplaceAllUsesWith(uint32_t before, uint32_t after);
+
  private:
   std::unique_ptr<Module> module_;
+  spvtools::MessageConsumer consumer_;
```

Comment:
This is causing -Werror build failures because consumer_ is never used.  Perhaps add it later when you have uses of it?

Reply:
It is being used.  It is returned by the "consumer()" member function, which is used to pass the consumer on to the def-use manager when it is being built (ir_context.h line 144).

Reply:
Looking at it closer, it was "consumer_" in the def use manger that was causing the problem.  The uses of it were moved to the IRContext.

I've removed it.

Reply:
Ah, yes.  It was the one in DefUseManager.  Thanks.